### PR TITLE
Allow null values to be manipulated at runtime

### LINKED ISSUE #1036: ddlog-sql: NullPointerException when trying to insert null for varchar column value
Repro diff on DDlog commit 0ccf6394
```
diff --git a/sql/src/test/java/ddlog/JooqProviderTest.java b/sql/src/test/java/ddlog/JooqProviderTest.java
index 300490f2..511831f9 100644
--- a/sql/src/test/java/ddlog/JooqProviderTest.java
+++ b/sql/src/test/java/ddlog/JooqProviderTest.java
@@ -64,7 +64,7 @@ public class JooqProviderTest {

     @BeforeClass
     public static void setup() throws IOException, DDlogException {
-        String s1 = "create table hosts (id varchar(36) with (primary_key = true), capacity integer, up boolean)";
+        String s1 = "create table hosts (id varchar(36) with (primary_key = true), capacity integer, up boolean, tester varchar(36))";
         String v2 = "create view hostsv as select distinct * from hosts";
         String v1 = "create view good_hosts as select distinct * from hosts where capacity < 10";
         String checkArrayParse = "create table junk (testCol integer array)";
@@ -136,10 +136,10 @@ public class JooqProviderTest {
     public void testSqlOpsWithBindings() {
         // Insert statements.
         create.insertInto(table("hosts"))
-              .values("n1", 10, true)
+              .values("n1", 10, true, null)
               .execute();
-        create.batch(create.insertInto(table("hosts")).values("n54", 18, false),
-                     create.insertInto(table("hosts")).values("n9", 2, true))
+        create.batch(create.insertInto(table("hosts")).values("n54", 18, false, null),
+                     create.insertInto(table("hosts")).values("n9", 2, true, null))
               .execute();
```

Partial stack trace of error:
```
java.lang.NullPointerException
	at org.jooq_3.13.1.DEFAULT.debug(Unknown Source)
	at ddlogapi.DDlogRecord.checkNull(DDlogRecord.java:26)
	at ddlogapi.DDlogRecord.<init>(DDlogRecord.java:95)
	at com.vmware.ddlog.DDlogJooqProvider.toValue(DDlogJooqProvider.java:511)
	at com.vmware.ddlog.DDlogJooqProvider.access$700(DDlogJooqProvider.java:86)
	at com.vmware.ddlog.DDlogJooqProvider$QueryVisitor.visitInsert(DDlogJooqProvider.java:266)
	at com.vmware.ddlog.DDlogJooqProvider$QueryVisitor.visit(DDlogJooqProvider.java:214)
	at com.vmware.ddlog.DDlogJooqProvider$QueryVisitor.visit(DDlogJooqProvider.java:200)
	at org.apache.calcite.sql.SqlCall.accept(SqlCall.java:139)
	at com.vmware.ddlog.DDlogJooqProvider.execute(DDlogJooqProvider.java:146)
	at org.jooq.tools.jdbc.MockStatement.execute0(MockStatement.java:191)
	at org.jooq.tools.jdbc.MockStatement.execute(MockStatement.java:267)
	at org.jooq.tools.jdbc.MockStatement.execute(MockStatement.java:262)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:209)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:453)
	at org.jooq.impl.AbstractDMLQuery.execute(AbstractDMLQuery.java:897)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:371)
	at org.jooq.impl.AbstractDelegatingQuery.execute(AbstractDelegatingQuery.java:119)
	at ddlog.JooqProviderTest.testSqlOpsWithBindings(JooqProviderTest.java:137)
```

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Fixes #1036

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
@amytai hopefully this will fix the issue.

**Comment by amytai** [comment]
Can we also
1) Add a test to check what happens to SQL without bindings (parsing literals)
2) Add a test to see that this gets rejected on the DDlog (?) side if you try to insert `null` into a `not null` column, as per Mihai's [question](https://github.com/vmware/differential-datalog/pull/1037#pullrequestreview-721741113).

File: sql/src/test/java/ddlog/JooqProviderTest.java [hunk 1]
```diff
@@ -129,6 +129,14 @@ public void testSqlOpsNoBindings() {
         assertFalse(goodHostsResultsAfterDelete.contains(test3));
     }
 
+    @Test
+    public void testInsertNull() {
+        // Issue 1036
+        create.insertInto(table("hosts"))
+                .values("n1", 10, null)
```
Comment (lalithsuresh) [inline sql/src/test/java/ddlog/JooqProviderTest.java:136]: Can you also add test cases for reads, deletes and updates? They have different code paths (that should eventually terminate in toValue() or maybeOption(), but it'd be still good to check.

**Comment by mbudiu-vmw** [comment]
Additional testing has uncovered additional bugs, e.g., #1041, which has been fixed in a separate PR.
@amytai

File: sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java [hunk 2]
```diff
@@ -164,14 +180,22 @@ public DDlogJooqProvider(final DDlogAPI dDlogAPI, final List<String> sqlStatemen
 
     private void rollback() {
         try {
+            if (trace)
+                System.out.println("Rolling back transaction");
             dDlogAPI.transactionRollback();
+            if (trace)
+                System.out.println("Transaction rolled back");
         } catch (final DDlogException e) {
+            if (trace)
```
Comment (lalithsuresh) [inline sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java:189]: Okay for now but we should be using sl4j instead of all the if(trace) -> stdout stuff.

**Review by lalithsuresh — APPROVED** [review]

**Comment by lalithsuresh** [comment]
LGTM

File: sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java (hunk 2 above)
Comment (mbudiu-vmw) [reply sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java:189]: I don't know if logging makes sense for testing.
I actually used this to debug a nasty bug where the exception was not caught here and the next test would fail because the started transaction was not rolledback.
